## Programming part of Homework 2 (Data Structures, Fall 2024)

## Name: Chun-i Chen
## Student ID Number: 112590022

### Programming problem 1
Consider the problem of finding the binary representation of a positive decimal integer Problem 1. Please use Python to implement the pseudo-code you provide for solving the problem. You should provide the iterative and recursive functions named with `Dec_to_Bin_Ite()` and `Dec_to_Bin_Rec()` respectively. Then, perform a comparison on these two versions by measuring the execution time. Besides, we also consider the number of recursive calls made in the recursive approaches. Hence, you need to
1. implement the iterative approach as a function named as `Dec_to_Bin_Ite()`,
2. provide the recursive version for the approach named as `Dec_to_Bin_Rec()`, and
3. compare these two function with the same input in terms of running time and write what you have observed.

We will use an in-built python library `timeit` and the module function `timeit.timeit()` for measuring the running time respective. In you observation, please indicate the ways to measure the time and this will be provided in the template.

**Note:** For the programming problems, please do some experiments by yourself to observe the running time and the number of recursive calls. You need to report what you have observed from the experiments when you submit your homework.

In [34]:
def Dec_to_Bin_Ite(n):
    binary = ''

    while(n):
        b = n%2
        n = n//2
        binary = ('1' if b==1 else '0' ) + binary
    return binary
    
print(Dec_to_Bin_Ite(10))
print(Dec_to_Bin_Ite(19))
print(Dec_to_Bin_Ite(63))

1010
10011
111111


In [33]:
def Dec_to_Bin_Rec(n):
    if n==0: return ''
    else: return Dec_to_Bin_Rec(n//2) + str(n%2)

print(Dec_to_Bin_Rec(10))
print(Dec_to_Bin_Rec(19))
print(Dec_to_Bin_Rec(63))

1010
10011
111111


In [43]:
#---------------------------------------------------------timeit
import timeit as ti
n=5625
#---------------------------------------------------------Method 1
print("Running time comparison (Timer Method 1):")

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    Dec_to_Bin_Ite(n)
tend = ti.default_timer() #timer end計時器關閉
t_ite=tend - tstart#開始與結束的差值

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    Dec_to_Bin_Rec(n)
tend = ti.default_timer() #timer end計時器關閉
t_rec=tend - tstart#開始與結束的差值

print("Iterative on ", n)
print("iterative approach:", t_ite)
print("Recursive approach:", t_rec)

#---------------------------------------------------------Method 2
print("Running time comparison (Timer Method 2):")

def iterative_test():
    Dec_to_Bin_Ite(n)
    
def recursive_test():
    Dec_to_Bin_Rec(n)
    
t_ite = ti.timeit("iterative_test()", setup="from __main__ import iterative_test", number = 10000)#timeit(函數名稱,執行次數)
t_rec = ti.timeit("recursive_test()", setup="from __main__ import recursive_test", number = 10000)

print("Iterative on ", n)
print("iterative approach:", t_ite)
print("Recursive approach:", t_rec)


Running time comparison (Timer Method 1):
Iterative on  5625
iterative approach: 0.017834999998740386
Recursive approach: 0.024668300000485033
Running time comparison (Timer Method 2):
Iterative on  5625
iterative approach: 0.015575299999909475
Recursive approach: 0.02437090000603348


#### Please write down the observations you have after the comparisons:
##### Inputs (values of n) for five experiments
1. 1050<br>
2. 2625<br>
3. 3675<br>
4. 4725<br>
5. 5625<br>
##### Results of comparisons
*******************************************************************************************************
**Method 1**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Iteration      | 0.014493900002 | 0.021322100074 | 0.018662500079 | 0.016664700001 | 0.025666999980 |
|Recursion      | 0.021456099989 | 0.029938800028 | 0.024456500076 | 0.024464700022 | 0.025554699940 |

**Method 2**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Iteration      | 0.012669099960 | 0.015030100010 | 0.015862900065 | 0.016367999953 | 0.015925899963 |
|Recursion      | 0.020288800006 | 0.022842300008 | 0.028425100027 | 0.025355900056 | 0.032285499969 |

*******************************************************************************************************

##### Observations
After the above five experiments, it was found that there have a few times where the recursion time is faster than the iteration time with the same input, and both of them are very close. However, in most cases, the iteration time is significantly faster than the recursion time, I believe this is due to the overhead of function calls, which causes some delay.
<hr>

### Programming problem 2
A ***permutation*** $P$ of size $n$ is a way of arranging the numbers between 1 and $n$, where each number must be used and can only be used once. For example, 5 4 1 2 3 is a permutation of 1 2 3 4 5 where $P(5)=1$, $P(4)=2$, $P(1)=3$, $P(2)=4$, and $P(3)=5$. An ***inversion*** in a permutation $P$ is a pair of numbers $i \mbox{ and} j \in\{1, \ldots, n\}$ with $i<j$ such that the larger number appears to the left of the smaller one in the permutation, that is $P(j)<P(i)$. In the example, the pair of (3, 5) is an inversion in $P$. The ***inversion number*** of a permutation is the total number of inversions. Please write the Python fuctions mentioned in Problem 4 and compare the performance in term of running time. The things to be provided are as below.
1. Implement the iterative approach using Python and name the function as `inversion_number_iterative()`.
2. Please implement the linear recursive version you provide in Problem 4(c) with the function name of `inversion\_number\_recursive()`.
3. Write the binary recursive version you provide in Problem 4(d) with the function named as `inversion\_number\_tworecurs()`.
4. Compare these functions with the same input in terms of running time and write what you have observed.

In [18]:
def inversion_number_iterative(P):
    count = 0
    for i in range(len(P)):
        for j in range(i+1,len(P)):
            if(P[i]>P[j]): count+=1
    return count

perm=[4, 6, 2, 5, 1, 3]
# perm = [12, 11, 13, 5, 6, 7]
# perm=[35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82]
print('The inversion number of ', perm, 'is', inversion_number_iterative(perm))

The inversion number of  [4, 6, 2, 5, 1, 3] is 10


In [22]:
def inversion_number_recurs(P,i):
    if(i==len(P)-1): return 0
    total = 0
    for j in range(i+1,len(P)):
        if(P[i]>P[j]): total+=1
    return total + inversion_number_recurs(P,i+1)

def inversion_number_recursive(P):
    return inversion_number_recurs(P,0)    
perm=[4, 6, 2, 5, 1, 3]
#perm = [12, 11, 13, 5, 6, 7]
# perm=[35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82]
print('The inversion number of ', perm, 'is', inversion_number_recursive(perm))

The inversion number of  [4, 6, 2, 5, 1, 3] is 10


In [19]:
def counting(P,left,right):
    total = 0
    mid = (left+right)//2
    
    for i in range(left,mid+1,1):
        for j in range(right,mid,-1):
            if(P[i]>P[j]): total+=1
    #print(f'i={i} j={j} counting: {total}')
    return total


def inversion_number_tworecursive(P,left,right):
    if(left+1 < right and right<len(P)):
        mid = (left+right)//2
        total = inversion_number_tworecursive(P,left,mid) # left
        total += inversion_number_tworecursive(P,mid+1,right) # right
        total += counting(P,left,right)
        #print(total)
        return total
    elif(left+1 == right):
        #print(f'{left},{right}')
        return 1 if(P[left]>P[right]) else 0
    return 0


def inversion_number_tworecurs(P):
    return inversion_number_tworecursive(P,0,len(P)-1)

perm=[4, 6, 2, 5, 1, 3]
# perm = [12, 11, 13, 5, 6, 7]
# perm=[35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82]
print('The inversion number of ', perm, 'is', inversion_number_tworecurs(perm))

The inversion number of  [4, 6, 2, 5, 1, 3] is 10


In [31]:
#---------------------------------------------------------timeit
import timeit as ti

perm = [6154, 8124, 1254, 3012, 5469, 4512, 4321, 4720]
# perm=[35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82]

#---------------------------------------------------------Method 1
print("Running time comparison (Timer Method 1):")

tstart = ti.default_timer() #timer start 計時器開啟

for i in range(10000):
    inversion_number_iterative(perm)
tend = ti.default_timer() #timer end計時器關閉
t_ite=tend - tstart#開始與結束的差值

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    inversion_number_recursive(perm)
tend = ti.default_timer() #timer end計時器關閉
t_rec=tend - tstart#開始與結束的差值

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    inversion_number_tworecurs(perm)
tend = ti.default_timer() #timer end計時器關閉
t_twr=tend - tstart#開始與結束的差值

print("Permutation:", perm)
print('The inversion number of is', inversion_number_iterative(perm),'(iterat)')
print('The inversion number of is', inversion_number_recursive(perm),'(linear)')
print('The inversion number of is', inversion_number_tworecurs(perm),'(binary)')

print("Iterative approach:", t_ite)
print("Recursive approach:", t_rec)
print("Binrecurs approach:", t_twr)

#---------------------------------------------------------Method 2
print("Running time comparison (Timer Method 2):")

def iterative_test():
    inversion_number_iterative(perm)

def recursive_test():
    inversion_number_recursive(perm)

def binrecurs_test():
    inversion_number_tworecurs(perm)
    
t_ite = ti.timeit("iterative_test()", setup="from __main__ import iterative_test", number = 10000)#timeit(函數名稱,執行次數)
t_rec = ti.timeit("recursive_test()", setup="from __main__ import recursive_test", number = 10000)
t_twr = ti.timeit("binrecurs_test()", setup="from __main__ import binrecurs_test", number = 10000)

print("Iterative approach:", t_ite)
print("Recursive approach:", t_rec)
print("Binrecurs approach:", t_twr)

Running time comparison (Timer Method 1):
Permutation: [6154, 8124, 1254, 3012, 5469, 4512, 4321, 4720]
The inversion number of is 16 (iterat)
The inversion number of is 16 (linear)
The inversion number of is 16 (binary)
Iterative approach: 0.0363361999989138
Recursive approach: 0.033943800000997726
Binrecurs approach: 0.043609500004095025
Running time comparison (Timer Method 2):
Iterative approach: 0.02981790000194451
Recursive approach: 0.03100270000140881
Binrecurs approach: 0.03969039999356028


#### Please write down the observations you have after the comparisons:
##### Inputs for five experiments
1. [35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82]<br>
2. [12, 11, 13, 5, 6, 7]<br>
3. [33, 55, 66, 77, 88, 99, 22]<br>
4. [36, 52, 78, 94, 24, 67, 512, 2, 589, 166, 664, 556, 13, 54, 612]<br>
5. [6154, 8124, 1254, 3012, 5469, 4512, 4321, 4720]<br>
##### Results of comparisoms
*******************************************************************************************************
**Method 1**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Iteration      | 0.063402499996 | 0.020801999999 | 0.024288499997 | 0.075522699997 | 0.036336199999 |
|Recursion      | 0.056478999999 | 0.026973799999 | 0.026437799999 | 0.086334000000 | 0.033943800001 |
|Binrecurs      | 0.092417200001 | 0.033945800002 | 0.036876900005 | 0.118061000001 | 0.043609500004 |

**Method 2**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Iteration      | 0.046621599999 | 0.017474599998 | 0.020512500007 | 0.067353999999 | 0.029817900002 |
|Recursion      | 0.054773399999 | 0.023375900004 | 0.024586200001 | 0.081205499999 | 0.031002700001 |
|Binrecurs      | 0.093508499995 | 0.033836500006 | 0.036545499999 | 0.117666099999 | 0.039690399994 |

*******************************************************************************************************

##### Observations
According to the results, the iteration time is faster than the recursion time, even though the theoretical time complexity of binary recursion is O(nlogn), which should be faster than the other two, experimentally recursion is still much faster than binary recursion. Binary recursion requires significantly more function calls than regular recursion, so I believe this is due to the delay caused by function calls, as mentioned earlier in Program I.
<hr>